### ISSUE TITLE
Automatic (catalog, cached tiles, ..) update after data change

### ISSUE THREAD

**Issue by iferencik** [urbica/martin#288]
Hello there

I'd like to know if  setting WATCH_MODE=true makes martin reload the config file.
I am publishing many postgis layer with even more properties/columns and in order to limit these i need to use a configuration file.
I am thinking to create a python script that will generate the config yaml and it would be really nice if martin could re-read this in WATCH_MODE

thank you for this great software box

**Comment by sharkAndshark** [comment]
If this is on road map,can i try this?

**Comment by stepankuzmin** [comment]
Hi @sharkAndshark,

Sure!

**Comment by sharkAndshark** [comment]
> Hi @sharkAndshark,
> 
> Sure!

Based on the discusion in #349 ,there will be no watch mode? @stepankuzmin

**Comment by sharkAndshark** [comment]
Based on the discusion in https://github.com/maplibre/martin/issues/349 ,there will be no watch mode? @stepankuzmin

**Comment by nyurik** [comment]
@sharkAndshark I think we may need to rethink the watch mode, and **possibly** add it back before the next release.

* Refresh should NOT be initiated on GET request, as this is a direct path to DDOSing the service, plus it may get cached by some proxy in between, or the browser
* Martin should be configurable with regular refresh intervals, i.e. every 10 minutes
* Martin should be configurable to allow a POST request to refresh -- easier to filter with firewall rules, does not get cached by proxies.  Configuration should allow a subset of IPs, and only from the localhost by default.
* rework refresh using [shared mutable state](https://actix.rs/docs/application/#shared-mutable-state) concepts.  I think it should be something like "get X without the shared multi-threaded state, unless X does not exist, in which case get a `read` of a read/write lock, and copy the latest state to the worker's state". Not certain if this is doable though.
* Add support for volatile/versioned data sources:
   * tilejson should return a slightly different path for the actual tiles. E.g. if the path is `/rpc/getmvt.json`, the tile path should be `/rpc/getmvt/<version>/<z>/<x>/<y>.pbf`, where the `<version>` is 1,2,...  This path should be dynamic, and change whenever a new version is detected.  Only applies when it is possible to access different versions at the same time.  Note that this approach would only work if there is only one instance of Martin, which is not scalable.
   * if the backend is an mbtiles (sqlite file) or a PMtiles file, the file could in theory be renamed while the pooled connections still have it open, and a new file is placed in the same location -- so this could work with versions.

That said, I really feel like the more substantial changes should be done with kubernetes and helm and deployments, and the `watch` mode is a development hack, making infrastructure too fragile.  A better set up would be:
* set up a kubernetes+helm deployment with multiple redundant instances of Martin
* when a new config/data sources are needed, create a new helm deployment version with new configuration (could even automate these steps)
* initiate a gradual roll-out of the new version, using automated tests to ensure the roll-out is going on ok, and rolling back to the last stable configuration if it is not

**Comment by sharkAndshark** [comment]
@nyurik I will give it a a shot and then more discussion

**Comment by nyurik** [comment]
Sounds good! @stepankuzmin any thoughts?

**Comment by stepankuzmin** [comment]
@sharkAndshark sure, that would be great 👍

**Comment by pascal-codetaal** [comment]
@nyurik We are currently using the latest version of martin, and there are some real benefits in using this version.
But the only thing that is stil a downside in the new version, is the remove of the watch or the option to reload the index.
We are running on a Kubernetes cluster and we reload the index by creating a new pod, and killing the other one.
But this process can take up to 50 seconds or so, that's the time the user has to wait before, before being able to see the updated data. 
So is there a possibility to add another way to refresh the data? 
Perhaps we can try to help with implementing this feature?

**Comment by nyurik** [comment]
@pascal-codetaal I'm beginning to think we really do need to reimplement it, and some help would be great. So lets have a dedicated `POST /reload` endpoint.

My only concern is that we would want to avoid a global lock on each request. I'm not yet certain how to do that best.

**Comment by iferencik** [comment]
> @nyurik We are currently using the latest version of martin, and there are some real benefits in using this version. But the only thing that is stil a downside in the new version, is the remove of the watch or the option to reload the index. We are running on a Kubernetes cluster and we reload the index by creating a new pod, and killing the other one. But this process can take up to 50 seconds or so, that's the time the user has to wait before, before being able to see the updated data. So is there a possibility to add another way to refresh the data? Perhaps we can try to help with implementing this feature?

@pascal-codetaal 
if you do a 
```
    kubectl rollout restart deployment <depl_name> 

```

there is no downtime. kube will ensure the old pod(s) is/are working while the new one gets deployed.

Just make sure you have enough capacity on the node to spin a new deployment. 

Also, you should ideally have more(>1) replicas.

I have the same issues and being under time pressure I resorted to something else
I created a sh [script] (https://github.com/UNDP-Data/martin-docker/blob/main/reactive_entrypoint.sh) that checks  the local config config file against a remote one located at a web accessible url (Azure file share).

An env vars specify how often the url/config is pulled and where the URL is pulled from.

This might not be very elegant but it **works like a charm** specially because i use a [tool](https://github.com/UNDP-Data/martin-config) to create (optionally upload to azure) the config  files. We actually update the layers on a daily, sometimes hourly basis. In future i will pool the remote config once a day.


The tool has a quirk that it creates configs only for tables that have the "publish=True" in the table [description](https://github.com/UNDP-Data/martin-config/blob/main/src/martin_config/db.py#L333). The same is valid for columns, if you mark a column with [publish=False](https://github.com/UNDP-Data/martin-config/blob/main/src/martin_config/db.py#L309) the toll will skip it.
This is not a very good idea and I think I will scrape it in favor of a more robust technique like  revoking read rights for the user under which martin connects to the database.

**Comment by pascal-codetaal** [comment]
We don't have a downtime, it just can take up to 50 seconds before the new pod is up and running...
We let users upload data to our db and they don't like to wait an extra 50 seconds before they can see the updated data. I'm going to check your script, thx

**Comment by saosebastiao** [comment]
I would love to be able to reload the schema config without a full restart. The PostgREST project has two cool ways to do this:

https://postgrest.org/en/stable/schema_cache.html#schema-cache-reloading

* listen for a SIGUSR1 signal
* Using Postgres' LISTEN/NOTIFY with a dedicated channel

The second method in particular is interesting because it allows the DB admin to set up custom triggers that would send notifications to the custom channel any time a geospatial table changes.

**Comment by nyurik** [comment]
There are currently two issues I see (i'm sure there are more):
* Performance - if we introduce a global lock for each request in order to access a singleton configuration, we might not be able to scale efficiently. This might be negligible - will need to evaluate in some simple stress test (e.g. tiny actix app that returns a static string after accessing some global lock var - while under a massive stress test, and the app is limited to just several CPUs)
* Source ID stability - when reloading, the auto-generated IDs could end up with a different one, e.g. in case multiple sources have the same table name. Possible solutions:
  * add an ability to generate a hash param from an internal "long ID": `hash(server_name, server_port, db_name, namespace, table, geo_column)`, and users can set it with `id_format: '{table}.{hash}'` in the `auto_publish` settings.
  * add some heuristics to try to match existing IDs to the new IDs based on the internal "long ID"

**Comment by tordans** [comment]
We just ran into this issue after switching from pg_tileserv to Martin. With pg_tileserv all new tables where discovered right away. Now, we have to restart our Martin Docker container to get an updated table list.

Our docker setup https://github.com/FixMyBerlin/atlas-geo/blob/main/docker-compose.yml starts our backend server with db, processing and tiles. It looks like we don't have a good way to restart the tile docker _after_ the app image closed and finished processing our data…

Is there some API like `<marinserver.com>/refresh` that we could ping to notify Martin to refresh tables?

**Comment by nyurik** [comment]
@tordans we all want that api to happen - I plan to work on it in the future, but no certainty on the timeframe -- it is mostly a personal project, so if anyone has any time to work on it, I will be happy to guide and assist that effort.

**Comment by pascal-codetaal** [comment]
Hello @nyurik and the Martin contributors,

We've been closely following the discussion about the need for a watch mode or similar feature to enable dynamic config reloading in Martin. As we understand the significance of this feature for users, and us,  who need to see updated data without substantial downtime, we're keen on contributing to its development.

Though we're not currently well-versed in Rust, we're ready to put in the necessary time to become proficient and contribute meaningfully to this project. We're particularly interested in understanding the preferred approach.

Could you provide us with some guidance on where you believe we should start, or if there are any particular areas of the codebase or documentation that we should familiarize ourselves with first? 

We're looking forward to collaborating with the community on this feature. Thank you for considering our offer to help.

**Comment by nyurik** [comment]
@pascal-codetaal I will be happy to help you.

The very first thing I think we should evaluate is the performance aspect outlined in https://github.com/maplibre/martin/issues/288#issuecomment-1494739125 -- create a tiny Actix app using one of the examples they have as a starting point. Inside that app, create two endpoints, e.g. `/lock` and `/nolock` to see the impact of the global lock.  You can use the `oha` tool (used in the justfile) to compare the performance of both endpoints.

This should get you familiarized with the Rust ecosystem, making it much easier to make changes to a bigger project.

```rust
#[derive(Clone)]
struct State {
    pub values: HashMap<String, String>,
    pub values_mutex: Arc<Mutex<HashMap<String, String>>>,
}

impl State {
    pub fn new() -> Self {
        let values = vec![(String::from("key"), String::from("value"))]
            .into_iter()
            .collect::<HashMap<String, String>>();
        Self {
            values_mutex: Arc::new(Mutex::new(values.clone())),
            values,
        }
    }
}

#[get("/lock")]
async fn get_lock(state: Data<State>) -> String {
    let vals = state.values_mutex.lock().unwrap();
    vals.get("key").unwrap().clone()
}

#[get("/nolock")]
async fn get_nolock(state: Data<State>) -> String {
    state.values.get("key").unwrap().clone()
}
```

**Comment by pascal-codetaal** [comment]
@nyurik the development will be done by @mesudBe-Orbit 
He will start with the development of the separate Actix application, and he will push this to a public repository.

**Comment by mesudBe-Orbit** [comment]
@nyurik @pascal-codetaal the development is complete, and you can access the tiny application and view the results by following this link: https://github.com/mesudBe-Orbit/actix-lock-stress-tester?tab=readme-ov-file#test-results
In short, under these conditions, it seems there is no significant difference.

**Comment by nyurik** [comment]
awesome news, thanks!  So this should make it easier.

The current config system process:
* In `start()`, read config file if given, using current env vars for any variable expansions.
* Merge together CLI parameters and the config file (as well as any env vars)
* Run `.resolve()` to match the configuration to the actual state of the database/files, and create a `ServerState` (tile sources, fonts, etc)
https://github.com/maplibre/martin/blob/a864b9981c9b8cfeaf3b4a90d40ec2635674bd59/martin/src/bin/martin.rs#L15-L27

The `resolve` method creates cache and an ID Resolver, and uses `resolve_tile_sources` to process all tile sources **in parallel**.     

https://github.com/maplibre/martin/blob/a864b9981c9b8cfeaf3b4a90d40ec2635674bd59/martin/src/config.rs#L115

`IdResolver` maps "unique full source IDs" (like `database.namespace.tablename` or even longer) to some short name the user may want, like `tablename`.  If the short name already exists, it will append numbers to it, e.g. `tablename.1`, but if the same full unique name is passed to IdResolver, it will re-return the same short name as before - so my point above about stable resolution is mostly done - same table will get the same short name.

PostgreSQL resolution is much more involved than pmtiles/mbtiles. It runs these steps in parallel for all configured postgres connections:
* create a `PgBuilder` instance from the `PgConfig` instance.
* in parallel, query PG to discover all existing tables and functions that can be used as sources (e.g. `query_available_tables`)
* for all configured tables/funcs, match them with the ones discovered in the database, and create a new source instance - in parallel (e.g. `instantiate_tables`)
* if auto-discovery is enabled, instantiate the rest of the tables/funcs that have not been matched up with the configuration

https://github.com/maplibre/martin/blob/a864b9981c9b8cfeaf3b4a90d40ec2635674bd59/martin/src/pg/config.rs#L116

So one path to  implement watch mode:
* Change all endpoint functions to accept `sources: Data<RwLock<TileSources>>` instead of `Data<TileSources>`, and get a read lock on all of them. Note that `RwLock` might not be good for our usecase, but it should get us started.
* after parsing and resolving `Config` struct, store it (as  `Arc<Mutex<Config>>`) inside the actix's `Data<...>`
* add a `/refresh` endpoint that gets `Data<Arc<Mutex<Config>>>` from Actix
* refresh should call `resolve` to compute new set of sources, and once done, lock and update all tile sources using the write of the RwLock.
* the `Config::resolve` should be changed to preserve IdResolver instance between calls - e.g. move it to constructor?

**Comment by nyurik** [comment]
> @mesudBe-Orbit @pascal-codetaal the development is complete, and you can access the tiny application and view the results by following this link: https://github.com/mesudBe-Orbit/actix-lock-stress-tester?tab=readme-ov-file#test-results In short, under these conditions, it seems there is no significant difference.

Please run your code with `cargo run --release`, and also you should probably increase the number of requests, e.g. aim for about 5 minute run (?)

**Comment by pascal-codetaal** [comment]
@nyurik the 5 minute run had the same result for both endpoints.
We started with the implementation of  changing the endpoint functions. 

> Change all endpoint functions to accept sources: Data<RwLock<TileSources>> instead of Data<TileSources>, and get a read lock on all of them. Note that RwLock might not be good for our usecase, but it should get us started.

**Comment by nyurik** [comment]
@pascal-codetaal sounds good. One other thing - I would recommend creating a Martin fork, create a branch there, and create a `draft` pull request into martin.  Make sure to commit and push your code often (even if it does not compile).  This way everyone can see the progress, and make suggestions as you go, without waiting for the full completion, and to correct mistakes early on.  This way you will also get the benefit of constant CI checks.

**Comment by mesudBe-Orbit** [comment]
@nyurik we have created a draft PR into martin. [you can find it here](https://github.com/maplibre/martin/pull/1179) And commit and push our latest code (the latest commit was about mistakenly added repetition).

**Comment by nyurik** [comment]
@mesudBe-Orbit yes, thanks, i saw it and already commented on that PR. Keep adding things to it (by pushing to the same branch), and once it is ready to be merged, click the `Ready to review` button.  Make sure to either merge main branch into it, or rebase on main and force push.

**Comment by nyurik** [comment]
pmtiles support for this would be good as well.  I outlined how it can be done in https://github.com/stadiamaps/pmtiles-rs/issues/40 - and once done, we can use it in Martin too.

**Comment by xyzhuo** [comment]
Thanks.

**Comment by RevTpark** [comment]
Are there updates on this issue/feat request and/or do we have a approx timeline when can we expect it to be done?

**Comment by CommanderStorm** [comment]
We have done some work to make the internal catalog paralell-updatable with little-no performance impact in https://github.com/maplibre/martin/pull/1585.

What we still need is another PR adding configuration anauto-updating behaviour
Basically, cherry pick some of the changes from https://github.com/maplibre/martin/pull/1179 and rebase them with changes onto master => adressing https://github.com/maplibre/martin/pull/1179#pullrequestreview-2149936800.

**Comment by CommanderStorm** [comment]
In terms of timeline, we don't provide deadlines as this is a volunteer driven project.

I would be happy to review a PR if you want to speed this up this way ^^

**Comment by gabeschine** [comment]
Would a PR that listens to a signal (such as `SIGHUP`) and reloads the config be welcome?

**Comment by nyurik** [comment]
certainly!  Be farewarned, there is a reason it hasn't been done for a long time - it is a bit complex. The actual method for refreshing is less of a problem, it's the internal plumbing to update the config while the service is already running

**Comment by gabeschine** [comment]
> certainly! Be farewarned, there is a reason it hasn't been done for a long time - it is a bit complex. The actual method for refreshing is less of a problem, it's the internal plumbing to update the config while the service is already running

Thank you - I was mis-understanding this comment, apparently:

> We have done some work to make the internal catalog paralell-updatable with little-no performance impact in https://github.com/maplibre/martin/pull/1585.

I'll take a look, though. I do have a production use-case that will eventually require a live config re-load (really, just re-listing paths from a static config). One option is a rolling restart of all production dynos, but that feels like a big hammer if a `SIGHUP` approach is available.

**Comment by Auspicus** [comment]
@gabeschine how did you go with this? I also have a production use-case where this will be necessary either via `SIGHUP` but preferrably `POST /reload`. If this was available via HTTP there would need to be some consideration of authorization/authentication.

**Comment by nyurik** [comment]
@CommanderStorm do you think watch mode support can be added to the existing config structure without any breaking changes? I can't think of any, but you may have thoughts on that

**Comment by CommanderStorm** [comment]
Yes. No reason that this is breaking.
I won't block 1.0.0 on this.

Not sure if `POST /reload` or `SIGHUP` is needed.
Maybe for postgres, but all file based sources have better APIs to watch for changes

- For file based sources (mbtiles, cog, sprites, styles, fonts), we can just get [notified](https://docs.rs/notify/latest/notify/) for changes.
  
  This is cheap, reliable and correct. There is no reason not to do this that I can see.
- For postgres can POLL for changes every n seconds
- PMtiles is considered static since we don't currently have autoconfiguration, but if we had we also would need to refres every X seconds

In terms of config, the following options work:

- <details><summary>centralised config</summary>
  <p>
  
  ```yaml
  refresh:
    # This will reload postgres sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    postgres: 10s
    
    # enables POST /reload if the following BEARER token is present
    # This will reload all sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    enable_with_bearer: ABCEDF
    
    # This signal will reload all sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    enable_with_signal: SIGHUP
  ```
  
  </p>
  </details> 

- <details><summary>decentralised config</summary>
  <p>
  
  ```yaml
  refresh:
    # enables POST /reload if the following BEARER token is present
    # This will reload all sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    enable_with_bearer: ABCEDF
    # This signal will reload all sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    enable_with_signal: SIGHUP
  
  
  postgres:
    # This will reload postgres sources in the following way:
    # - new sources get added
    # - dead sources get removed and their cache get whiped
    # - changed soures get their cache whiped
    refresh_every: 10s
  ```
  
  </p>
  </details> 

> [!TIP]
> In terms of implementation, there is no obvious "A has to come before B" on this one.
> 
> Picking a subsystem and implementing the reloading just for that should be quite doable.
> Code wise, splitting into "what changes happend (add, remove, edit)" and applying changes seems sensible, but open to different designs 😉.
> For some resources/tile sources edit might also not make entire sense, this is tbd.
> 
> The code for the configuration lives here under the new architecture
> https://github.com/maplibre/martin/tree/main/martin/src/config/file

**Comment by gabeschine** [comment]
> [@gabeschine](https://github.com/gabeschine) how did you go with this? I also have a production use-case where this will be necessary either via `SIGHUP` but preferrably `POST /reload`. If this was available via HTTP there would need to be some consideration of authorization/authentication.

My apologies for missing this comment. I just updated my notification settings in GitHub so hopefully that doesn't happen again. I have made no progress on this. In the medium-term, we use Heroku for serving and a rolling restart of the Martin server instances there is a reasonable path forward. As you indicated above, though, there's an inherent delay, and it's a big hammer. I am motivated to avoid using this big hammer, although I can't promise exactly when I'll be contributing a PR to move us closer to it.

**Comment by CommanderStorm** [comment]
I have done a bit of thinking and I think the following achritecture could work.
-> Comments?

## RFC: A Decentralized Reloading Architecture

Instead of a single coordinator, this design uses multiple, independent reloader actors. Each reloader is specialized for a single source type.

This approach presents a new challenge: ensuring that concurrent updates from different reloaders do not corrupt the shared source catalog. The solution is to make the `TileSourceManager` responsible for serializing these concurrent partial updates and giving out sensible ids for each source in the catalog via the `IdResolver`.

```mermaid
flowchart TB
 subgraph TSM["TileSourceManager"]
        TSM_DS("DashMap
        SourceID -> BoxedSource")
        TSM_C["TileCache
        SourceID, XYZ, Query -> Bytes, Meta"]
  end
 subgraph reloaders["Decentralized Reloaders"]
        PGR["PostgresReloader"]
        MBR["MBTilesReloader"]
        PMR["PMTilesReloader"]
        COGR["COGReloader"]
  end
    A["main()"] -. spawns .-> PGR & MBR & PMR & COGR
    TSM -- updates after applying updates --> PC["public catalog"]
    PGR -- updates --> TSM
    MBR -- updates --> TSM
    PMR -- updates --> TSM
    COGR -- updates --> TSM
```

### Key Roles

*   **`TileSourceManager` (Concurrent State Manager):** The single source of truth for the live tile sources and caches.
    It is responsible for ensuring that concurrent updates from different reloaders are applied strongly consistent (assuming the same id allocation for both) regarding IDs and eventually consistent regarding the state of the sources in one update.
    The public catalog is only updated after the additions/edits from the `[Type]Reloader` is fully applied.
*   **`[Type]Reloader` Actors (e.g., `PostgresReloader`):** These are self-contained, long-running tasks. Each reloader:
    1.  **Maintains its own state** of the sources it manages (e.g., a `BTreeSet<String>` of table names for `PostgresReloader`, or a `BTreeMap<String, u64>` of file paths and modification times for `MBTilesReloader`).
    2.  **Detects** changes in its domain (polling or file watching).
    3.  **Discovers** the new set of sources and their versions.
    4.  **Constructs** a `ReloadAdvisory` by comparing the newly discovered sources with its internal state.
    5.  **Calls** the `TileSourceManager` to apply the changes.
    6.  On successful application, **updates its internal state**.
    7.  **Updates** the public catalog with the new sources.

### Data Flow for a Reload Operation (Example: Postgres)

```mermaid
sequenceDiagram
    participant PGR as PostgresReloader
    participant DB as PostgreSQL
    participant TSM as TileSourceManager

    loop Every `refresh_every` seconds
        PGR->>DB: Discover tables & functions
        DB-->>PGR: Return `next_ids`
        PGR->>PGR: Get internal state (`previous_ids`)
        PGR->>PGR: advisory = ReloadAdvisory.from_sets(previous_ids, next_ids, fn_instantiate)
        PGR->>TSM: apply_changes(advisory)
        TSM-->>PGR: Done
        PGR->>PGR: Update internal state to `next_ids`
    end
```

## Phase 1: `TileSourceManager` and `ReloadAdvisory`

This phase defines the shared structures. The `ReloadAdvisory` is enhanced to contain the diffing logic, which is called by the reloaders. The `TileSourceManager` simply applies these changes, relying on the underlying thread-safety of `DashMap`.
Its primary purpose is to give out IDs in a consistent manner, ensuring that the order of IDs is predictable and stable.

Ids should be done via `IdResolver`:
- additions always get a new ID
- updates keep the ID (-> if data on disk changes, this should not affect the ID but does the cache)
- removals "burn" IDs

```rust
use std::collections::{BTreeSet, BTreeMap};
use std::sync::Arc;
use dashmap::DashMap;
use martin_core::source::Source;
use martin_core::tiles::TileCache;

pub struct DeletedSource {
    // public, primary key
    pub name: String,
    // private, secondary key.
    // influences ONLY the id renaming of `name` if there is a conflict
    pub unique_name: String,
}

pub struct NewSource {
    // public, primary key
    pub name: String,
    // private, secondary key.
    // influences ONLY the id renaming of `name` if there is a conflict
    pub unique_name: String,
    pub source: Box<dyn Source>,
}

#[derive(Default)]
pub struct ReloadAdvisory {
    pub additions: Vec<NewSource>,
    pub updates: Vec<NewSource>,
    pub removals: BTreeSet<DeletedSource>,
}

impl ReloadAdvisory {
    /// Generates an advisory for unversioned sources (e.g., Postgres, PMTiles).
    /// 
    /// Any change is treated as a removal and an addition.
    pub fn from_sets<F>(
        previous_ids: &BTreeSet<String>,
        next_ids: &BTreeSet<String>,
        mut instantiator: F,
    ) -> Self
    where
        F: FnMut(&String) -> Box<dyn Source>,
    {
        let source_ids_to_remove = previous_ids.difference(next_ids).cloned().collect();
        let sources_to_add = next_ids.difference(previous_ids).map(|id| NewSource {
            id: id.clone(),
            source: instantiator(id),
        }).collect();

        Self { sources_to_add, source_ids_to_remove, sources_to_update: Vec::new() }
    }

    /// Generates an advisory for versioned sources (e.g., MBTiles, COG).
    /// 
    /// It can distinguish between additions, removals, and updates (version changes).
    pub fn from_maps<F, V: Eq + Copy>(
        previous_map: &BTreeMap<String, V>,
        next_map: &BTreeMap<String, V>,
        mut instantiator: F,
    ) -> Self
    where
        F: FnMut(&String) -> Box<dyn Source>,
    {
        let mut advisory = Self::default();
        for id in previous_map.keys() {
            if !next_map.contains_key(id) {
                advisory.source_ids_to_remove.insert(id.clone());
            }
        }

        for (id, &next_version) in next_map {
            match previous_map.get(id) {
                Some(&prev_version) if next_version != prev_version => {
                    advisory.sources_to_update.push(NewSource {
                        id: id.clone(),
                        source: instantiator(id),
                    });
                }
                None => {
                    advisory.sources_to_add.push(NewSource {
                        id: id.clone(),
                        source: instantiator(id),
                    });
                }
                _ => {} // Unchanged
            }
        }
        advisory
    }
}

/// Manages the live set of sources and their caches.
/// 
/// A broad lock is not needed because each reloader manages a mutually exclusive
/// set of sources, and `DashMap` provides atomic per-key operations.
/// TileCache and IdResolver have an inner Arc, so the whole TileSourceManager is Cheap to clone.
pub struct TileSourceManager {
    tile_sources: Arc<DashMap<String, Box<dyn Source>>>,
    tile_cache: TileCache,
    // take a cookie, neo -> https://youtu.be/MQJlBpPLHUM
    the_oracle: IdResolver,
}

impl TileSourceManager {
    pub fn new(cache_size_mb: usize) -> Self {
        Self {
            tile_sources: Default::default(),
            tile_cache: TileCache::new(cache_size_mb),
        }
    }
    
    pub async fn apply_changes(&self, advisory: ReloadAdvisory) {
        // updates may be time critical -> do them first
        for new_source in advisory.updates {
            let id = &self.the_oracle.resolve(&new_source.name, new_source.unique_name);
            self.tile_cache.invalidate_source(&id);
            self.tile_sources.insert(id, new_source.source);
        }
        // Prevent stale data reads
        self.tile_cache.run_maintenance().await;  

        // additions -> make them available
        for new_source in advisory.additions {
            let id = &self.the_oracle.resolve(&new_source.name, new_source.unique_name);
            self.tile_sources.insert(id, new_source.source);
        }
      
        // all sources are now online, update the public catalog
        self.update_public_catalog(advisory).await;

        // Removals -> not time criticical, just GC
        for deleted_source in &advisory.removes {
            let id = &self.the_oracle.resolve(&deleted_source.name, deleted_source.unique_name);
            self.tile_sources.remove(&id);
            self.tile_cache.invalidate_source(id);
        }
    }
}
```

## Phase 2: Implement the Decentralized Reloaders

This phase involves creating the independent reloader actors, each with its own internal state.

-   **`PostgresReloader`**:
    -   Maintains an internal `BTreeSet<String>` of active source IDs.
    -   On a polling tick, it discovers the new list of sources from the DB.
    -   It calls `ReloadAdvisory::from_sets()` to compute the diff.
    -   It calls `TileSourceManager::apply_changes` and afterwards updates its internal set.

-   **`MBTilesReloader`**, **`COGReloader`**:
    -   Maintains an internal `BTreeMap<String, u64>` mapping file paths to modification timestamps.
    -   On a file watcher event, it re-scans its paths and gets new modification times.
    -   It calls `ReloadAdvisory::from_maps()` to compute the diff.
    -   It calls `apply_changes` and afterwards updates its internal map.

-   **`PMTilesReloader`**:
    -   PMTiles files are immutable, so content changes are not expected. This reloader will be similar to `PostgresReloader`, watching for file additions and removals.
    -   Maintains an internal `BTreeSet<String>` of active file paths.
    -   Uses `ReloadAdvisory::from_sets()` for diffing.

## Phase 3: Configuration

The configuration remains unchanged, but its usage is now decentralized.
Each reloader actor during its initialization gets its part of the configuration to determine if it should be enabled and how it should behave.

```yaml
# Used by PostgresReloader
postgres:
  connections: ...
  auto_discovery:
    refresh_every: 15s

# Used by MBTilesReloader, etc.
mbtiles:
  paths: [...]
  watch: true
```

## Rollout Strategy

1.  **Implement Phase 1:** Define the `TileSourceManager` with its update lock and the `ReloadAdvisory` with its `from_sets`/`from_maps` constructors.
2.  **Implement Phase 2:** Create the individual reloader actors, each with its own internal state and detection logic.
3.  **Implement Phase 3:** Wire up the configuration to enable and configure each reloader independently.
4.  **Documentation:** Update the project documentation to explain the new reloading features and configuration.

**Comment by gabeschine** [comment]
> I have done a bit of thinking and I think the following achritecture could work. -> Comments?

Awesome thank you for putting this together! I am of limited value because I'm still learning the architecture. Please take these comments with a healthy grain of salt.

> ## RFC: A Decentralized Reloading Architecture
> Instead of a single coordinator, this design uses multiple, independent reloader actors. Each reloader is specialized for a single source type.
> 
> This approach presents a new challenge: ensuring that concurrent updates from different reloaders do not corrupt the shared source catalog. The solution is to make the `TileSourceManager` responsible for serializing these concurrent partial updates and giving out sensible ids for each source in the catalog via the `IdResolver`.

I am not sure if this is helpful or not, but, it could be possible to assign sources a `uuid` every time they are resolved or have their version bumped, and then maintain an alias map from `public name -> uuid` that would represent the public catalog entry-point. Using the `uuid` for cache keying would virtually eliminate the possibility of cache pollution.

It adds more indirection, so maybe not a great idea. Just throwing it out there.

>   * PMTiles files are immutable, so content changes are not expected. This reloader will be similar to `PostgresReloader`, watching for file additions and removals.
>   * Maintains an internal `BTreeSet<String>` of active file paths.
>   * Uses `ReloadAdvisory::from_sets()` for diffing.

A PMTiles file can be overwritten, which would need to be treated as a version bump. S3, for example, will atomically swap objects, and computes ETags automatically.

> ## Phase 3: Configuration
> The configuration remains unchanged, but its usage is now decentralized. Each reloader actor during its initialization gets its part of the configuration to determine if it should be enabled and how it should behave.
> 
> # Used by PostgresReloader
> postgres:
>   connections: ...
>   auto_discovery:
>     refresh_every: 15s

Does `auto_discovery` relate to `auto_publish`? I'm assuming not, but those two terms are close enough in name to me. Can save naming for a PR of course, but wanted to flag that confusion.

> # Used by MBTilesReloader, etc.
> mbtiles:
>   paths: [...]
>   watch: true

Would you be open to using a consistent configuration terminology/syntax across all source types?

**Other questions**

- How should we reconcile the refresh op with the new `on_invalid` policies? Ideas:
  - `on_invalid: warn`: only add or version bump existing sources, never remove (ie, optimistic policy).
  - `on_invalid: abort`: do not refresh at all if there are any warnings, or, use same policy as `on_invalid: warn`.
    - ^ this could get complicated since the proposal above de-centralizes the refresh and does not coordinate it between source types
- Does this proposal formalize the concept of a source version? For debugging I can imagine that being useful: if the source + loaded version are exposed and available in the metrics, or from the web UI / JSON catalog, it would help an eng-ops person discover if a failed load is responsible for stale data at a glance. This comment is related to the above about using `uuids`, at least tangentially.

Let me know if there are any ways I can be more useful here.

**Comment by CommanderStorm** [comment]
> A PMTiles file can be overwritten, which would need to be treated as a version bump. S3, for example, will atomically swap objects, and computes ETags automatically.

Yea, I know. We have a bug there btw:
- https://github.com/stadiamaps/pmtiles-rs/issues/40

> > ## Phase 3: Configuration
> > The configuration remains unchanged, but its usage is now decentralized. Each reloader actor during its initialization gets its part of the configuration to determine if it should be enabled and how it should behave.
> > # Used by PostgresReloader
> > postgres:
> > connections: ...
> > auto_discovery:
> > refresh_every: 15s
> 
> Does `auto_discovery` relate to `auto_publish`? I'm assuming not, but those two terms are close enough in name to me. Can save naming for a PR of course, but wanted to flag that confusion.

I meant `auto_publish`, but not sure if this is too naiive.
If you configure your tables manually, one can still appear and we would not pick it up, so this naiive approache likely does not work -> I don't know how this should be configured.

> Would you be open to using a consistent configuration terminology/syntax across all source types?

Yes, obviously.
To be transparent: I have not done deep thinking on how this should be configured and mostly been focusing on how to integrate this into the rest of the system 😉

**Comment by gabeschine** [comment]
> > A PMTiles file can be overwritten, which would need to be treated as a version bump. S3, for example, will atomically swap objects, and computes ETags automatically.
> 
> Yea, I know. We have a bug there btw:
> 
> * [Handle backend data changes stadiamaps/pmtiles-rs#40](https://github.com/stadiamaps/pmtiles-rs/issues/40)

Copy that. Does this issue block a complete solution for this RFC? If my read is correct, the implications are that so long as the Martin cache is disabled, and the underlying PMTiles does not change its metadata in a meaningful way, a refreshed PMTiles file will work transparently. On the flip side, any change in metadata (bounds, zoom range, etc) would require a Martin reload until this issue is resolved. Sound right?

> > Does `auto_discovery` relate to `auto_publish`? I'm assuming not, but those two terms are close enough in name to me. Can save naming for a PR of course, but wanted to flag that confusion.
> 
> I meant `auto_publish`, but not sure if this is too naive. If you configure your tables manually, one can still appear and we would not pick it up, so this naive approachelikely does not work -> I don't know how this should be configured.

With `on_invalid: warn` and `auto_publish` off, an initial startup could skip a configured but non-existent table source from Postgres (or other source). We'd want that table to get picked up once it becomes valid. I think the refresh configuration needs to be divorced from `auto_publish` entirely. Thoughts?

> > Would you be open to using a consistent configuration terminology/syntax across all source types?
> 
> Yes, obviously. To be transparent: I have not done deep thinking on how this should be configured and mostly been focusing on how to integrate this into the rest of the system 😉

SG - will ignore all of that for now! Thanks.

**Comment by CommanderStorm** [comment]
I think for pmtiles it should be resonably simple to add other features like
- https://github.com/stadiamaps/pmtiles-rs/issues/40
- #2180 

I don't want to promise new features for now, since I have not looked into it.
Since those likely need more involved tests, I would argue to push those changes off to contributors who need those changes, since they are also fairly rewarding. 😉

**Comment by CommanderStorm** [comment]
@nyurik might it be possible for you to review https://github.com/maplibre/martin/issues/288#issuecomment-3707778253 over the next few weeks?

I have a bit of time on an airplane, where I can work on things like this.

**Comment by iboates** [comment]
Hi I just discovered Martin recently and I mistakenly thought that it would automatically detect new schemas and tables as they get created automatically. Apparently this is not the case. I found this ticket, and it is a lot to take in, but it seems by my understanding to be about making this possible, i.e. the catalog will very quickly reflect creation of new tables in the database.

Am I correct in my understanding?

**Comment by CommanderStorm** [comment]
Yes, this issue is for enabling automatic catalog/cache/.. updates.
I would love a design review on the design that I proposed here to solve this -> https://github.com/maplibre/martin/issues/288#issuecomment-3707778253

It is a bit complicated, but not sure where to simplify it.

**Comment by Auspicus** [comment]
I'm just gonna throw my two cents in here after playing around with implementing the RFC @CommanderStorm described: https://github.com/maplibre/martin/issues/288#issuecomment-3707778253.

It appears there are multiple different interpretations of what watch mode or reload should do, and different deployments might require different things:

### Changes to config.yaml

**What does this mean?**

A change was made to the config.yaml file Martin used to start.

**How could this be achieved?**

Considering that config.yaml controls a wide range of things (number of workers, cache size, data sources, path prefix, etc), it seems untenable to do anything other than a full STOP THE WORLD restart. At which point, as @nyurik said, it's probably better to let the configured restart policy apply (via K8s, snaps config, etc). Perhaps configuration + behavior could be added to restart Martin when this file is changed?

**Notes:**
- Arguments passed via the command line are merged with the config from the config.yaml file. How does this get reconciled when the config.yaml file is changed but the arguments are not?
- If `--save-config` is passed then what should happen? A restart loop might happen if we're not careful.

### Changes to data sources referenced via args or config

**What does this mean?**

A change (modification or removal) was made to a connection already established by Martin (PostGIS database connection, PMTiles remote file, COG remote file, MBTiles local file), etc.

**How could this be achieved?**

Remote files or database connections might need to be polled. Local files can use the OS specific file watch events.

**Notes:**
- Notifying consumers of a cache change (ETags could help here)
- File connections like MBTiles, PMTiles or COG could be corrupted if modified while in use
- For named sources (eg. `my-source-name: ./file-path/source-123.mbtiles`) any modifications or deletions should be considered a fatal error as a failure to load these during startup will block startup (perhaps not with `on_invalid`)?

### Sources added or removed to folders that were configured for discovery

**What does this mean?**

A file was added or removed from a folder that Martin was configured to discover sources from. For example Martin was started with `martin ./my-tile-folder` and an MBTiles file was created at `./my-tiles-folder/new.mbtiles`.

**How could this be achieved?**

Again OS specific file watcher on the folder configured for discovery should be enough here. Additions and removals are valid. Modifications to existing files are probably still unsafe. Should probably be explicitly opt-in either at the connection level or top level...

**Notes:**
- A path accessed prior to the source being added might return a 404 depending on what cache headers were returned.

**Comment by henriquecolini** [comment]
Hey everyone,

I've been following this discussion for a bit now, and forgive me if I'm saying something stupid or obvious, but... what about storing a timestamp for each cached tile? And ignoring it if it's older than some configurable amount of time? That would solve 99% of the issue for me.

Am I missing something here?

**Comment by henriquecolini** [comment]
I took a look at the code and realized the tile cache uses `moka::future::Cache`, which has a `time_to_live` and `time_to_idle`, so the suggestion in my previous comment goes even simpler. I must be missing something, right? Because it would be really easy to add an `Option<Duration>` parameter to `TileCache::new` (or even create a `TileCache::with_ttl` as not to break the API) and an optional ttl field to config files.

**Comment by CommanderStorm** [comment]
@henriquecolini 
That is a different issue. You want this instead:
- https://github.com/maplibre/martin/issues/2505

Basically, that related issue just needs plumbing work, yes.

This issue is about "what if I have configured `martin some/dir/` and then `foo.mbtiles` spawns? Will it get picked up?"

**Comment by CommanderStorm** [comment]
@henriquecolini The configrable TTL/TTI support was merged and is avaliable in the latest release.

**Comment by jo-chemla** [comment]
Nice to hear TTL/TTI is now configurable!
For those getting here via a github search, the added blocks from the [config-file.yml](https://maplibre.org/martin/config-file/) are copied below, can be written in human-readable formats like 1h, 3600s, and default is no-expiry (meaning entries only get evicted by size pressure)

```yml
# Cache configuration
cache:
  # ...
  
  # Maximum lifetime for all cache entries (time-to-live from creation).
  # Entries are evicted after this duration regardless of access.
  # Supports human-readable formats: "1h", "30m", "1d", "3600s".
  # default: null (no expiry, entries only evicted by size pressure)
  expiry: null

  # Maximum idle time for all cache entries (time-to-idle since last access).
  # Entries are evicted if not accessed within this duration.
  # default: null (no idle timeout)
  idle_timeout: null

  # Tile-specific TTL override. Takes precedence over `cache.expiry` for tiles.
  # default: null (inherits from `cache.expiry`)
  tile_expiry: null

  # Tile-specific idle timeout override. Takes precedence over `cache.idle_timeout` for tiles.
  # default: null (inherits from `cache.idle_timeout`)
  tile_idle_timeout: null
```

**Comment by CommanderStorm** [comment]
- https://github.com/maplibre/martin/pull/2717

So we "now" (once windows CI is happy) have the ability to hot-reload mbtiles if they are configured using an directory and the directory content changes.

This is by now way the be-all, end-all of solutions but it is a great step into the right direction.

We still need the following migrated to the new architecture:
- PMTilesReloader
- COGReloader
- PostgresReloader

As always, volunteers welcome 😉 
At this point, I think this is just the thankless blumbing work that is left.
Should not be hard and is a good first issue

**Comment by CommanderStorm** [comment]
So we have merged a larger step towards this goal:
- [x] COG
- [x] PMTiles
- [x] MBTiles
- [ ] Postres
  - [x] Decoupling init and discovery
  - [x] Adding fingerprints
  - [ ] Refactoring the underlying traits we use to be better fitting
  - [ ] Add the postgres reloader
- [ ] DuckDB

So let's call it progress.
As always, many things ongoing.

<img width="648" height="500" alt="Image" src="https://github.com/user-attachments/assets/0073b41f-00ff-4c27-87da-3c9c97bd6599" />

**Comment by CommanderStorm** [comment]
Proposed architecture has been mostly implemented (+- a few traits for testing seams)
